File: python/task3c/S3_PIPE_03.py
# S3_PIPE_03 - String normalization pipeline

from functools import reduce
import re

def pipe(*fns):
    return lambda x: reduce(lambda v, fn: fn(v), fns, x)

# Pipeline steps
strip_ws = lambda s: s.strip()
lowercase = lambda s: s.lower()
collapse_spaces = lambda s: re.sub(r'\s+', ' ', s)

normalize = pipe(strip_ws, lowercase, collapse_spaces)

# Test
test_str = "  Ala   Ma   Kota  "
print(f"Input: '{test_str}'")
print(f"Output: '{normalize(test_str)}'")

File: js/task3c/S3_PIPE_03.js
// S3_PIPE_03 - String normalization pipeline

const pipe = (...fns) => x => fns.reduce((v, fn) => fn(v), x);

// individual transforms
const trim = s => s.trim();
const lowercase = s => s.toLowerCase();
const normalizeSpaces = s => s.replace(/\s+/g, " ");

// the pipeline
const normalizeString = pipe(trim, lowercase, normalizeSpaces);

// tests
console.log(normalizeString("  Hello   World  "));       // "hello world"
console.log(normalizeString("  LOTS    OF   SPACES  ")); // "lots of spaces"
console.log(normalizeString("Already Normal"));          // "already normal"
console.log(normalizeString("   "));                      // ""
